File: rust/executor/redirect.rs
use super::Executor;
use crate::parser::ast::Redirect;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResolvedRedirect {
    pub fd: i32,
    pub op: String,
    pub target: String,
}

pub async fn resolve_redirects(
    redirects: &[Redirect],
    executor: &Executor,
) -> Vec<ResolvedRedirect> {
    let mut resolved = Vec::new();
    for r in redirects {
        let target = executor.expand_word(&r.target).await;
        resolved.push(ResolvedRedirect {
            fd: r.fd,
            op: r.op.clone(),
            target,
        });
    }
    resolved
}

pub fn apply_input_redirect(
    redirects: &[ResolvedRedirect],
    stdin: &str,
    executor: &Executor,
) -> String {
    for r in redirects {
        if r.op == "<" {
            let path = executor.resolve_path(&r.target);
            let content = executor
                .bridge
                .fs_read_file(&path)
                .as_string()
                .unwrap_or_default();
            return content;
        }
        if r.op == "<<<" {
            return format!("{}\n", r.target);
        }
        if r.op == "<<" {
            // Heredoc — body would be in target
            return r.target.clone();
        }
    }
    stdin.to_string()
}

pub fn get_output_redirects(redirects: &[ResolvedRedirect]) -> Vec<ResolvedRedirect> {
    redirects
        .iter()
        .filter(|r| {
            r.op == ">"
                || r.op == ">>"
                || r.op == "&>"
                || r.op == "&>>"
        })
        .cloned()
        .collect()
}

File: src/executor/redirect.ts
import type { IFileSystem } from "@ambiently-work/mirage";
import type { Environment } from "../env/environment.js";
import type { Redirect } from "../parser/index.js";
import { expandWord, type SubExecFn } from "./expansion/index.js";

export interface ResolvedRedirect {
	fd: number;
	op: string;
	target: string;
}

export async function resolveRedirects(
	redirects: Redirect[],
	env: Environment,
	fs: IFileSystem,
	subExec: SubExecFn,
): Promise<ResolvedRedirect[]> {
	const resolved: ResolvedRedirect[] = [];

	for (const r of redirects) {
		const target = await expandWord(r.target, env, fs, subExec);

		resolved.push({
			fd: r.fd,
			op: r.op,
			target,
		});
	}

	return resolved;
}

export function applyInputRedirect(
	redirects: ResolvedRedirect[],
	currentStdin: string,
	fs: IFileSystem,
	resolvePath: (p: string) => string,
): string {
	let stdin = currentStdin;

	for (const r of redirects) {
		switch (r.op) {
			case "<": {
				const path = resolvePath(r.target);
				try {
					stdin = fs.readFile(path);
				} catch (e) {
					throw new Error(
						`${r.target}: ${e instanceof Error ? e.message : "No such file or directory"}`,
					);
				}
				break;
			}
			case "<<<": {
				// Here-string
				stdin = `${r.target}\n`;
				break;
			}
			case "<<": {
				// Here-document (body already extracted by parser)
				stdin = r.target;
				break;
			}
		}
	}

	return stdin;
}

export function getOutputRedirects(redirects: ResolvedRedirect[]): ResolvedRedirect[] {
	return redirects.filter(
		(r) =>
			r.op === ">" ||
			r.op === ">>" ||
			r.op === "&>" ||
			r.op === "&>>" ||
			r.op === ">&" ||
			r.op === "<>",
	);
}
